### PR TITLE
rootston: Cancel existing keyboard grab when changing focus

### PR DESCRIPTION
**Pull request by SirCmpwn**
This patch fixes issue #233. As described there, if you have two
windows open, then open a XDG popup on one, and left-click on the
other, the newly-clicked window will not receive keyboard input. This
is especially jarring as it still receives XDG activation.

The cause is that when trying to change our seat's focus, an existing
keyboard grab can ignore our attempt. So, we cancel existing grabs
before doing so. The commit message has lower-level details.

It's worth noting two alternative approaches that were discussed with
@acrisci on the IRC channel:

(A) Change each wlr_pointer_grab_interface's enter() implementation
to cancel itself and call wlr_seat_keyboard_notify_enter(), if it
expects to do nothing. However, it is messy to change every interface.

(B) Change wlroots to behave like weston, where a click on the new
window will neither set it active nor send it keyboard input. This
also involves changing every interface to return whether it is a
no-op, which is messy. And, this would be counter-intuitive behavior,
according to discussion on IRC and issue #233.

### CODE REVIEW

File: rootston/seat.c [hunk 1]
```diff
@@ -825,6 +825,11 @@ void roots_seat_set_focus(struct roots_seat *seat, struct roots_view *view) {
 	view_activate(view, true);
 	seat->has_focus = true;
 
+	// We want to unconditionally send keyboard input to the view we are
+	// focusing here, so cancel any existing grabs.
+	struct wlr_seat_keyboard_grab *curr_grab = seat->seat->keyboard_state.grab;
```
Comment (acrisci) [inline rootston/seat.c]: probably better to use `wlr_seat_keyboard_end_grab()`.

**Comment by acrisci** [comment]
Yeah, this is probably the most explicit and simple way to do it. The downside is it requires the library user to manually manage grabs in this case. It would be better if the grab could do this for you, but the downside of that is that now you have to explicitly have a rule of "all keyboard grabs shall end their grab when you send keyboard_notify_enter to another surface". That's probably safe to do though and I'm not necessarily against it.

**Comment by acrisci** [comment]
What's the status of this? Can I just commit the change I wanted and merge?

**Comment by SirCmpwn** [comment]
Sure.

**Pull request merged**